### ISSUE TITLE
Phantom edges appeared in concurrent update

### ISSUE THREAD

**Issue by YinAqu** [orientechnologies/orientdb#8608]
## Expected behavior  
No ghost edges exist

## Actual behavior  
Ghost edges exist. Run the test code bellow,  check the `_TeacherBlockingEdge`, you'll find out there are some ghost edges between `Teacher` and `Blocking`.

## Steps to reproduce  
Test code:
```
package issue;

import com.orientechnologies.orient.core.db.ODatabaseSession;
import com.orientechnologies.orient.core.db.ODatabaseType;
import com.orientechnologies.orient.core.db.OrientDB;
import com.orientechnologies.orient.core.db.OrientDBConfig;
import com.orientechnologies.orient.core.metadata.schema.OClass;
import com.orientechnologies.orient.core.metadata.schema.OType;
import com.orientechnologies.orient.core.record.ODirection;
import com.orientechnologies.orient.core.record.OVertex;
import com.orientechnologies.orient.core.sql.executor.OResultSet;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class Orient3ConcurrencyTest {

    private static final String DB_NAME = "test";

    private static final String TEACHER_STUDENT_EDGE = "_TeacherStudentEdge";

    private static final String TEACHER_BLOCKING_EDGE = "_TeacherBlockingEdge";

    private static final String TEACHER = "Teacher";

    private static final String TEACHER_PROP_NAME = "name";

    private static final String TEACHER_PROP_ID = "teacherId";

    private static final String STUDENT = "Student";

    private static final String STUDENT_PROP_TEACHER_ID = "teacherId";

    private static final String STUDENT_PROP_IDENTIFIER = "studentId";

    private static final String BLOCKING = "Blocking";

    private static final String BLOCKING_PROP_KEY = "key";

    // The final result should probably be only one Teacher record in database,
    // even more than one, the bocking edges final state is not correct.
    public static void main(String[] args) throws InterruptedException {
        OrientDB orientDB = new OrientDB("remote:localhost", "root", "123456", OrientDBConfig.defaultConfig());
        schema(orientDB);
        String teacherName = "Jim";
        // create one record first
        create(orientDB, UUID.randomUUID().toString(), teacherName);

        // concurrent create records
        ExecutorService executorService = Executors.newFixedThreadPool(30);
        for (int i = 0; i < 3000; i++) {
            executorService.submit(() -> create(orientDB, UUID.randomUUID().toString(), teacherName));
        }
        executorService.awaitTermination(2, TimeUnit.MINUTES);
        executorService.shutdown();
    }

    private static void create(OrientDB orientDB, String studentId, String teacherName) {
        ODatabaseSession db = orientDB.open("test", "admin", "admin");
        try {
            db.begin();
            String teacherId = UUID.randomUUID().toString();

            OVertex blockingV = getBlockingVIfNoneCreate(teacherName, db);

            Iterable<OVertex> candidateVs = blockingV.getVertices(ODirection.IN, TEACHER_BLOCKING_EDGE);
            Set<OVertex> candidates = new HashSet<>();
            candidateVs.forEach(candidates::add);
            Set<OVertex> ids = new HashSet<>();
            candidates.forEach(candidateV -> {
                Iterable<OVertex> studentVs = candidateV.getVertices(ODirection.OUT, TEACHER_STUDENT_EDGE);
                studentVs.forEach(ids::add);
                candidateV.delete();
            });

            OVertex teacherV = db.newVertex(TEACHER);
            teacherV.setProperty(TEACHER_PROP_NAME, teacherName);
            teacherV.setProperty(TEACHER_PROP_ID, teacherId);

            OVertex studentV = db.newVertex(STUDENT);
            studentV.setProperty(STUDENT_PROP_IDENTIFIER, studentId);
            studentV.setProperty(STUDENT_PROP_TEACHER_ID, teacherId);
            teacherV.addEdge(studentV, TEACHER_STUDENT_EDGE);

            ids.forEach(id -> {
                id.setProperty(STUDENT_PROP_TEACHER_ID, teacherId);
                teacherV.addEdge(id, TEACHER_STUDENT_EDGE);
            });

            teacherV.addEdge(blockingV, TEACHER_BLOCKING_EDGE);
            teacherV.save();
            db.commit();
        } catch (Exception e) {
            e.printStackTrace();
            db.rollback();
        } finally {
            db.close();
        }
    }

    private static OVertex getBlockingVIfNoneCreate(String name, ODatabaseSession db) {
        OResultSet rs = db.query("select from " + BLOCKING + " where " + BLOCKING_PROP_KEY + "=?", name);
        if (rs.hasNext()) {
            return rs.next().getVertex().orElseThrow(() -> new RuntimeException());
        } else {
            OVertex vertex = db.newVertex(BLOCKING);
            vertex.setProperty(BLOCKING_PROP_KEY, name);
            vertex.save();
            return vertex;
        }
    }

    private static void schema(OrientDB orientDB) throws InterruptedException {
        orientDB.createIfNotExists(DB_NAME, ODatabaseType.PLOCAL);

        try (ODatabaseSession db = orientDB.open(DB_NAME, "admin", "admin")) {
            OClass teacher = db.getClass(TEACHER);
            if (teacher == null) {
                teacher = db.createVertexClass(TEACHER);
                teacher.createProperty(TEACHER_PROP_ID, OType.STRING).setNotNull(true).createIndex(OClass.INDEX_TYPE.UNIQUE);
                teacher.createProperty(TEACHER_PROP_NAME, OType.STRING).setNotNull(true);
            }
            OClass student = db.getClass(STUDENT);
            if (student == null) {
                student = db.createVertexClass(STUDENT);
                student.createProperty(STUDENT_PROP_TEACHER_ID, OType.STRING).setNotNull(true);
                student.createProperty(STUDENT_PROP_IDENTIFIER, OType.STRING).setNotNull(true).createIndex(OClass.INDEX_TYPE.UNIQUE);
            }
            OClass blocking = db.getClass(BLOCKING);
            if (blocking == null) {
                blocking = db.createVertexClass(BLOCKING);
                blocking.createProperty(BLOCKING_PROP_KEY, OType.STRING).setNotNull(true).createIndex(OClass.INDEX_TYPE.UNIQUE);
            }
            if (db.getClass(TEACHER_STUDENT_EDGE) == null) {
                db.createEdgeClass(TEACHER_STUDENT_EDGE);
            }
            if (db.getClass(TEACHER_BLOCKING_EDGE) == null) {
                db.createEdgeClass(TEACHER_BLOCKING_EDGE);
            }
        }

    }

}

```

**Comment by markodjurovic** [comment]
Hi @YinAqu 
if you want to have exactly one BLOCKING vertex, I think that 
`private static OVertex getBlockingVIfNoneCreate(String name, ODatabaseSession db)`
should be synchronized.

**Comment by YinAqu** [comment]
About how to determine the ghost edges: run the test code above, assume there are 5 `Teacher` vertexes left at last,  run `select from Blocking` in BROWSE tab of Orientdb's web client, check the `_TeacherBlockingEdge` field, you'll find there are **5** edge's ids in this field; and then, go to the GRAPH tab,  run `select from Blocking` again, double click the Blocking vertex, only **one** edge will be displayed.

**Comment by YinAqu** [comment]
@markodjurovic
I never said I want a exactly one BLOCKING vertex result. If you had a careful understanding of the logic of the code, you would notice that the test code above creates one record first before concurrent creating, besides, there's a unique index on the `key` field of the BLOCKING.

What I said in comments is that the result should **probably** be one **`Teacher`** vertex, of course, there may be more than one `Teacher` vertex due to concurrent creating, that's not the issue. Actually, via the logic of the test code, each `Teacher` vertex must have a `_TeacherBlockingEdge` edge associated with the `Blocking` vertex. But in fact, assume I have 5 `Teacher` vertexes left as a result of running the test code, there is always only one of them having a `_TeacherBlockingEdge` edge, that's the issue. Where are the other 4 `_TeacherBlockingEdge` edges? In the logic of the test code, `_TeacherBlockingEdge` edge is never deleted directly, but through the delete operation of the `Teacher` vertex, which deletes associated edges as well. So, I think the problem is that a vertex delete operation in a transaction of OrientDB is not an atomic operation, it must delete every associated edges first, however, if an exception occurs after deleting the edge and before deleting the vertex, the vertex will survive.

About how to determine the ghost edges: run the test code, assume there are 5 `Teacher` vertexes left at last, execute `select from Blocking` in BROWSE tab of OrientDB's web client, check the `_TeacherBlockingEdge` field, you'll find there are 5 edge ids in this field. Then, go to the GRAPH tab, execute `select from Blocking` again, double click the Blocking vertex, only one edge will be displayed, where are the other 4 edges?

**Comment by YinAqu** [comment]
@markodjurovic
I never said I want a exactly one BLOCKING vertex result.  If you look closely at the logic of the code, you'll notice that the test code above first created a record before concurrent creation, besides, there's a unique index on the `key` field of the BLOCKING.

What I said in comments is that the result should **probably** have only one **`Teacher`** vertex left, of course, due to concurrent creation, there may be more than one `Teacher` vertex left, this is not the issue. Actually, by the logic of the test code, each `Teacher` vertex must have a `_TeacherBlockingEdge` edge associated with a `Blocking` vertex. But in fact, assuming that I have 5 `Teacher` vertexes left as a result of running the test code, there is always only one of them having a `_TeacherBlockingEdge` edge, which is the issue. Where are the other 4 `_TeacherBlockingEdge` edges? In the logic of the test code, `_TeacherBlockingEdge` edge will never be deleted directly, but through the delete operation of the `Teacher` vertex, which deletes associated edges as well. So, I think the problem is that a vertex delete operation in a transaction of OrientDB is not an atomic operation, it must first delete each associated edge, but if an exception occurs after the edges are deleted and before deleting the vertex, the vertex will survive.

About how to determine the ghost edges: run the test code, assuming that there are 5 `Teacher` vertexes left, execute `select from Blocking` in BROWSE tab of OrientDB's web client, check the `_TeacherBlockingEdge` field, you'll find there are 5 edge ids in this field. Then, go to the GRAPH tab, execute `select from Blocking` again, double-click the Blocking vertex, only one edge will be displayed, where are the other 4 edges?

**Comment by markodjurovic** [comment]
@YinAqu 
Thank you on detailed explanation, I will investigate this.

**Comment by markodjurovic** [comment]
Hello @YinAqu,
I found two potential issues. One is related to your application, and second to Graph tab in OrientDB studio.

For the first one I reproduced issue you have, and after I've added `studentV.save();` in create method, I've had desired behaviour. In other words each of Teacher vertices had an edge connecting it to Blocking vertex. Please respond if this works for you?

Second one is representation of graph in Graph tab of OrientDB studio. Although there were more than one Teacher vertices (and corresponding edges), when executing `select from Blocking` in view only one edge- vertex pair appeared. We will investigate this further.

**Comment by YinAqu** [comment]
@markodjurovic 
I don't think `studentV.save();` helps.
Actually executing `select from Teacher` in BROWSE tab you can always see that every Teacher vertex has an edge id in _TeacherBlockingEdge field, no matter adding  `studentV.save();` or not, but in Graph tab, only one Teacher vertex displays one _TeacherBlockingEdge by double-click, just like the Blocking vertex. In addition, you could verify this behavior with OrientDB's java api(in fact I verified only with OrientDB v2.2.34 java api when I tested this issue on it).

**Comment by YinAqu** [comment]
@markodjurovic 
For a pity, `studentV.save()` doesn't works for me. I don't know your "each of Teacher vertices had an edge connecting it to Blocking vertex" comes out by what clue, the _TeacherBlockingEdge field in BROWSE tab?

In my environment, executing `select from Teacher` in BROWSE tab I can always see that every Teacher vertex has an edge in _TeacherBlockingEdge field, but in Graph tab, only one Teacher vertex displays _TeacherBlockingEdge by double-clicking, no matter adding  `studentV.save()` or not. Additionally, the Teacher vertexes which lose their _TeacherBlockingEdge lose their _TeacherStudentEdge too, many Student vertexes which should be associated with one Teacher become orphans.

**Comment by markodjurovic** [comment]
Hi @YinAqu 
you are right. I was checking only with `select from Teacher` and `select from Blocking`.

**Comment by markodjurovic** [comment]
Hi @YinAqu ,

thank you for reporting this one. Fix will be included in 3.0.10
